src/check/canonicalize.zig

```diff
@@ -5532,62 +5522,67 @@ fn canonicalizeTypeApplication(self: *Self, apply: CIR.TypeAnno.Apply, parent_no
+                const alias_var = self.can_ir.pushTypeVar(self.can_ir.env.types.mkAlias(.{ .ident_idx = apply.symbol }, d.anno_var, arg_anno_slice), parent_node_idx, region);
 
-        // Create the main alias variable first
-        const alias_var = try self.can_ir.pushFreshTypeVar(parent_node_idx, region);
+                // Ensure slots exist for the alias and all its arguments
+                // Shrink the scratch var buffer now that our work is done
+                self.scratch_vars.clearFrom(scratch_anno_start);
 
-        // Create backing var immediately after it
-        _ = try self.can_ir.pushFreshTypeVar(parent_node_idx, region);
+                return alias_var;
+            },
+            .s_nominal_decl => |d| {
+                const header = self.can_ir.store.getTypeHeader(d.header);
+                const args = self.can_ir.store.sliceTypeAnnos(header.args);
 
-        // Create type argument variables sequentially after backing var
-        for (actual_args) |arg_idx| {
-            _ = try self.canonicalizeTypeAnnoToTypeVar(arg_idx);
-        }
+                // Check arity
+                if (args.len != actual_args.len) {
+                    return try self.can_ir.pushTypeVar(.err, parent_node_idx, region);
+                }
 
-        // Now set the alias content on the main variable
-        const alias_content = Content{
-            .alias = .{
-                .ident = .{ .ident_idx = apply.symbol },
-                .num_args = @intCast(actual_args.len),
-            },
-        };
+                const scratch_anno_start = self.scratch_vars.top();
+                for (actual_args) |arg_idx| {
+                    const arg_anno_var = try self.canonicalizeTypeAnnoToTypeVar(arg_idx);
+                    self.scratch_vars.append(self.can_ir.env.gpa, arg_anno_var);
+                }
+                const arg_anno_slice = self.scratch_vars.slice(scratch_anno_start, self.scratch_vars.top());
 
-        // Ensure slots exist for the alias and all its arguments
-        try self.can_ir.env.types.ensureAliasSlots(alias_var, @intCast(actual_args.len));
+                const nominal_var = self.can_ir.pushTypeVar(self.can_ir.env.types.mkNominal(
+                    types.TypeIdent{ .ident_idx = header.name },
+                    d.anno_var,
+                    arg_anno_slice,
+                    self.can_ir.env.idents.insert(self.can_ir.env.gpa, base.Ident.for_text(self.can_ir.module_name), Region.zero()),
+                ), parent_node_idx, region);
 
-        _ = self.can_ir.setTypeVarAt(@enumFromInt(@intFromEnum(alias_var)), alias_content);
+                // Shrink the scratch var buffer now that our work is done
+                self.scratch_vars.clearFrom(scratch_anno_start);
 
-        return alias_var;
+                return nominal_var;
+            },
+            else => @panic("scopeLookupTypeDecl only returns type declarations"),
```

Comment:
Should we have this be `unreachable` then? I believe that optimizes better in release builds (and still panics in debug builds)